Convert GraphQL schema to Protobuf

raminqaf opened the pull request:
Closes #4 
This pull request adds the logic that converts GraphQL schemas to Protobuf

torbsto commented on manager/src/main/java/com/bakdata/quick/manager/graphql/GraphQLToProtobufConverter.java:
```diff
@@ -0,0 +1,310 @@
+        final Builder descriptorProtoBuilder = DescriptorProto.newBuilder();
+        descriptorProtoBuilder.setName(graphQLObjectType.getName());
+
+        final FileDescriptorProto.Builder tempFile = FileDescriptorProto.newBuilder();
```
I don't see why you would need a tempFile. You can move the check to each corresponding addition

torbsto commented on manager/src/main/java/com/bakdata/quick/manager/graphql/GraphQLToProtobufConverter.java:
```diff
@@ -0,0 +1,310 @@
+     * @see com.google.protobuf.DescriptorProtos.FieldDescriptorProto.Type#TYPE_ENUM
+     * @see com.google.protobuf.DescriptorProtos.FieldDescriptorProto.Builder#setTypeName(String)
+     */
+    @NotNull
```
No need for `@NotNull`, we mark `@Nullable`

torbsto commented on manager/src/main/java/com/bakdata/quick/manager/graphql/GraphQLToProtobufConverter.java:
```diff
@@ -0,0 +1,310 @@
+        final List<EnumValueDescriptorProto> values = enumType.getDefinition()
+            .getEnumValueDefinitions()
+            .stream()
+            .map(enumValueDefinition -> EnumValueDescriptorProto.newBuilder()
```
I've just noticed that this is a bit tricky. Given the following GraphQL
```graphql
enum Foo {
  A, B, C
}

enum Bar {
  A, B, C
}
```
the resulting protobuf would be invalid, because the enum values would clash in the namespace.
You could prefix them with the type name.

torbsto commented on manager/src/test/java/com/bakdata/quick/manager/graphql/GraphQLToProtobufConverterTest.java:
```diff
@@ -0,0 +1,175 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class GraphQLToProtobufConverterTest {
```
:heart_eyes:

torbsto commented on manager/src/main/java/com/bakdata/quick/manager/graphql/GraphQLToProtobufConverter.java:
```diff
@@ -0,0 +1,310 @@
+    }
+
+    @VisibleForTesting
+    public GraphQLToProtobufConverter(final String protobufPackage) {
```
Well, you also could just use `new GraphQLToProobufConverter(new ProtobufConfig(...))`

torbsto commented on manager/src/main/java/com/bakdata/quick/manager/graphql/GraphQLConverter.java:
```diff
@@ -0,0 +1,59 @@
+import io.confluent.kafka.schemaregistry.ParsedSchema;
+
+/**
+ * An interface for converting GraphQL schema to a {@link ParsedSchema} object.
```
```suggestion
 * Converter for transforming a GraphQL schema to a {@link ParsedSchema} object.
```
Just as an example, similar to the comment above

torbsto reviewed (commented):
Thanks for the PR! Looks good overall, just a couple of nits and one open question regarding the conversion of enums.

raminqaf commented on manager/src/main/java/com/bakdata/quick/manager/graphql/GraphQLToProtobufConverter.java:
```diff
@@ -0,0 +1,310 @@
+            final Descriptor descriptor = this.convertToDescriptor(graphQLSchema);
+            return new ProtobufSchema(descriptor);
+        } catch (final DescriptorValidationException exception) {
+            throw new BadRequestException(
```
I am a bit confused about this comment. You are using the same exception in the Avro converter here: https://github.com/bakdata/quick/blob/ed00e6c8f164d66110030dd2cd9b13a5c5015dd1/manager/src/main/java/com/bakdata/quick/manager/graphql/GraphQLToAvroConverter.java#L160

raminqaf commented:
@torbsto Thanks for the review! I applied them to my code. here are some notes:
- Based on the [style guide of buf](https://docs.buf.build/lint/rules#enum_value_prefix), I prefixed all the enum fields with the type name. Moreover, I added an `UNSPECIFIED` field to the beginning of the enum descriptor.

- I refactored the Protobuf conversion tests, and the test style is unified with the tests in the Avro conversion.

- I removed the `tempFile` builder. I check if a message/enum exists before adding it to the file.

- I applied the Javadoc style you wished for and will keep this style in my future PRs.

- I replaced the `BadArgumentException` with an `InternalErrorException`, since the `DescriptorValidationErrorException` accours when do the conversion wrong.

torbsto commented on manager/src/main/java/com/bakdata/quick/manager/graphql/GraphQLToProtobufConverter.java:
```diff
@@ -0,0 +1,326 @@
+    }
+
+    private static Map<GraphQLScalarType, Type> scalarTypeMap() {
+        return Map.ofEntries(
```
Any particular reason you're using Map.ofEntries instead of Map.of?

torbsto commented on manager/src/main/java/com/bakdata/quick/manager/graphql/GraphQLToProtobufConverter.java:
```diff
@@ -0,0 +1,326 @@
+     * @see com.google.protobuf.DescriptorProtos.FieldDescriptorProto.Type#TYPE_ENUM
+     * @see com.google.protobuf.DescriptorProtos.FieldDescriptorProto.Builder#setTypeName(String)
+     */
+    private static FieldDescriptorProto buildFieldWithType(
```
```suggestion
    private static FieldDescriptorProto createFieldWithType(
```

All the methods returning a proto element start with `create...`, any reason this doesn't?

torbsto commented on manager/src/main/java/com/bakdata/quick/manager/graphql/GraphQLToProtobufConverter.java:
```diff
@@ -0,0 +1,326 @@
+
+    private static void handleObjectType(
+        final FileDescriptorProto.Builder fileBuilder,
+        final Builder currentMessage,
```
```suggestion
        final Descriptor.Builder currentMessage,
```
As a parameter, `Builder` is to generic

torbsto replied:
This is not resolved yet

torbsto replied:
This is not resolved yet

torbsto commented on manager/src/main/java/com/bakdata/quick/manager/graphql/GraphQLToProtobufConverter.java:
```diff
@@ -0,0 +1,326 @@
+        }
+    }
+
+    private static void handleObjectType(
```
Could you add JavaDoc here as well?

torbsto reviewed (commented):
I suggest we document what GraphQL schema elements Quick supports, both from a query and conversion perspective. I *think* unions and interfaces are the only elements Quick doesn't support. This can be done in a follow-up PR though

raminqaf commented on common/src/main/java/com/bakdata/quick/common/config/ProtobufConfig.java:
```diff
@@ -0,0 +1,57 @@
+        if (!NAMESPACE_PATTERN.matcher(protobufPackage).matches()) {
+            throw new ConfigurationException(
+                String.format(
+                    "The Protobuf package %s does not fulfill the naming convention of Protobuf specification.",
```
Not from Google directly but as a style guide from buf.

torbsto replied:
I think we shouldn't enforce a style guide

raminqaf replied:
Should I remove it then?

raminqaf commented on manager/src/main/java/com/bakdata/quick/manager/graphql/GraphQLToProtobufConverter.java:
```diff
@@ -0,0 +1,326 @@
+     * @return enum field name
+     */
+    private static String generateEnumFieldName(final String typeName, final String enumValueDefinitionName) {
+        return typeName.toUpperCase() + "_" + enumValueDefinitionName;
```
I think it is not a bad idea

torbsto commented on common/src/main/java/com/bakdata/quick/common/config/ProtobufConfig.java:
```diff
@@ -0,0 +1,43 @@
+@ConfigurationProperties(ProtobufConfig.PREFIX)
+@Getter
+public class ProtobufConfig {
+    private static final String PROTO_FILE_EXTENSION = ".proto";
```
It's just a coincidence though that the config suffix is the same as the proto file extension. To me, it's a bit confusing to have the extension variable here used for the prefix

torbsto commented on common/src/main/java/com/bakdata/quick/common/config/ProtobufConfig.java:
```diff
@@ -0,0 +1,43 @@
+ */
+@ConfigurationProperties(ProtobufConfig.PREFIX)
+@Getter
+public class ProtobufConfig {
```
I think there is no test for this class yet

raminqaf commented:
The only thing that bothers me is the `protobufPackage` in the `ProtobufConfig` class. I cannot change it to only `package` because it is a keyword in Java. So the config would look like this:
`quick.schema.protobuf.protobuf-package`. How about we change it to `packageName`?

torbsto commented:
You could do something like this:

```java
@ConfigurationInject
public ProtobufConfig(@Property(name = PREFIX + ".package") final String protobufPackage) {
    this.protobufPackage = protobufPackage;
}
```

But I'm also fine with `packageName`

torbsto commented on manager/src/main/java/com/bakdata/quick/manager/graphql/GraphQLConverter.java:
```diff
@@ -0,0 +1,59 @@
+        // extending the schema with an empty query type is necessary because parsing fails otherwise
+        final SchemaParser schemaParser = new SchemaParser();
+
+        // extending the schema with an empty query type is necessary because parsing fails otherwise
```
Comment is duplicated, see line 44

torbsto commented on manager/src/test/resources/schema/graphql/shouldConvertListType.graphql:
```diff
@@ -0,0 +1,11 @@
+    optionalComplexList: [ComplexObject]
+
+    requiredSimpleList: [Int]!
+    requiredComplexList: [ComplexObject]!
```
Can you add a test for `[Object!]` and `[Object!]!`?

torbsto replied:
Can you add a test for `[Object!]` and `[Object!]!`?
